pkg/snapshot/clonetree_pci.go

```diff
@@ -0,0 +1,135 @@
+//
+// Use and distribution licensed under the Apache license version 2.
+//
+// See the COPYING file in the root project directory for full text.
+//
+
+package snapshot
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	pciaddr "github.com/jaypipes/ghw/pkg/pci/address"
+)
+
+const (
+	// root directory: entry point to start scanning the PCI forest
+	// warning: don't use the context package here, this means not even the linuxpath package.
+	// TODO(fromani) remove the path duplication
+	sysBusPCIDir = "/sys/bus/pci/devices"
+)
+
+// ExpectedClonePCIContent return a slice of glob patterns which represent the pseudofiles
+// ghw cares about, pertaining to PCI devices only.
+// Beware: the content is host-specific, because the PCI topology is host-dependent and unpredictable.
+func ExpectedClonePCIContent() []string {
+	var fileSpecs []string
+	pciRoots := []string{
+		sysBusPCIDir,
+	}
+	for {
+		if len(pciRoots) == 0 {
+			break
+		}
+		pciRoot := pciRoots[0]
+		pciRoots = pciRoots[1:]
+		specs, roots := scanPCIDeviceRoot(pciRoot)
+		pciRoots = append(pciRoots, roots...)
+		fileSpecs = append(fileSpecs, specs...)
+	}
+	return fileSpecs
+}
+
+// scanPCIDeviceRoot reports a slice of glob patterns which represent the pseudofiles
+// ghw cares about pertaining to all the PCI devices connected to the bus connected from the
+// given root; usually (but not always) a CPU packages has 1+ PCI(e) roots, forming the first
+// level; more PCI bridges are (usually) attached to this level, creating deep nested trees.
+// hence we need to scan all possible roots, to make sure not to miss important devices.
+func scanPCIDeviceRoot(root string) (fileSpecs []string, pciRoots []string) {
+	trace("scanning PCI device root %q\n", root)
+
+	perDevEntries := []string{
+		"class",
+		"device",
+		"irq",
+		"local_cpulist",
+		"modalias",
+		"numa_node",
+		"revision",
+		"vendor",
+	}
+	entries, err := ioutil.ReadDir(root)
+	if err != nil {
+		return []string{}, []string{}
+	}
+	for _, entry := range entries {
+		entryName := entry.Name()
+		if addr := pciaddr.FromString(entryName); addr == nil {
+			// doesn't look like a entry we care about
+			continue
+		}
+
+		entryPath := filepath.Join(root, entryName)
+		pciEntry := findPCIEntryFromPath(root, entryName)
+		trace("PCI entry is %q\n", pciEntry)
+		fileSpecs = append(fileSpecs, entryPath)
+		for _, perNetEntry := range perDevEntries {
+			fileSpecs = append(fileSpecs, filepath.Join(pciEntry, perNetEntry))
+		}
+
+		if isPCIBridge(entryPath) {
+			trace("adding new PCI root %q\n", entryName)
+			pciRoots = append(pciRoots, pciEntry)
+		}
+	}
+	return fileSpecs, pciRoots
+}
+
+func findPCIEntryFromPath(root, entryName string) string {
+	entryPath := filepath.Join(root, entryName)
+	fi, err := os.Lstat(entryPath)
+	if err != nil {
+		trace("stat(%s) failed: %v\n", entryPath, err)
+		return entryPath // what else we can do?
```

Comment:
we could return err here and above on line 75 we could `continue`... that way we don't bury the error here (even though we're `trace()`ing it.

Reply:
this is probably a better direction, yes. Let me try & test this out a bit on a couple systems - I want to run against more complex PCI topologies.